proto/substrait/algebra.proto

```diff
@@ -478,47 +478,136 @@ message Expression {
+    // number of arguments specified in the function definition, and the
+    // argument types must also match exactly:
+    //
+    //  - Value arguments must be bound using FunctionArgument.value, and
+    //    the expression in that must yield a value of a type that a function
+    //    overload is defined for.
+    //  - Type arguments must be bound using FunctionArgument.type, and a
+    //    function overload must be defined for that type.
+    //  - Required enum arguments must be bound using FunctionArgument.enum
+    //    followed by Enum.specified, with a string that case-insensitively
+    //    matches one of the allowed options.
+    //  - Optional enum arguments must be bound using FunctionArgument.enum
+    //    followed by either Enum.specified or Enum.unspecified. If specified,
+    //    the string must case-insensitively match one of the allowed options.
+    repeated FunctionArgument arguments = 9;
+
+    // Must be set to the return type of the function, exactly as derived
+    // using the declaration in the extension.
+    Type output_type = 7;
+
+    // TODO: I'm not sure I understand what this does.
+    AggregationPhase phase = 6;
+
+    // If specified, the rows that are part of the window defined by
+    // upper_bound and lower_bound are ordered according to this list
+    // before they are aggregated. The first sort field has the highest
+    // priority; only if a sort field determines two rows to be equivalent is
+    // the next field queried. This field is optional, and is only allowed if
+    // the window function is defined to support sorting.
     repeated SortField sorts = 3;
-    Bound upper_bound = 4;
+
+    // TODO: I'm not sure I understand what these are for.
+    repeated Expression partitions = 2;
+
+    // Defines the row relative to the current row from which the window
+    // extends. The bound is inclusive. If the lower bound indexes a row
+    // greater than the upper bound, TODO (null range/no rows passed?
+    // wrapping around as if lower/upper were swapped? error? null?).
+    // Optional; defaults to the start of the partition.
     Bound lower_bound = 5;
-    AggregationPhase phase = 6;
-    Type output_type = 7;
-    repeated FunctionArgument arguments = 9;
 
-    // deprecated; use args instead
+    // Defines the row relative to the current row up to which the window
+    // extends. The bound is inclusive. If the upper bound indexes a row
+    // less than the lower bound, TODO (null range/no rows passed?
+    // wrapping around as if lower/upper were swapped? error? null?).
+    // Optional; defaults to the end of the partition.
+    Bound upper_bound = 4;
+
+    // Deprecated; use arguments instead.
     repeated Expression args = 8 [deprecated = true];
 
+    // Defines one of the two boundaries for the window of a window function.
     message Bound {
+      // Defines that the bound extends this far back from the current row.
```

Comment:
(style) I would probably use record here and elsewhere as opposed to row. Row tends to suggest something tabular in nature. We may have nested-structs, etc so I tend to use record instead of row to describe that.

Reply:
I copied the term used on the website, so I'll change both.

Reply:
thanks

Reply:
I couldn't find it on the website after all aside from `CurrentRow`, so I guess that's what had me confused. Replaced in the new comments.